### ISSUE TITLE
This condition will always return 'false' since the types have no overlap

### ISSUE THREAD

**Issue by josdejong** [microsoft/TypeScript#41379]
**TypeScript Version:**  4.0.5, 4.2.0-dev.20201103

**Search Terms:**

"This condition will always return 'false' since the types have no overlap", "conditional equality"

**Code**

```ts
let text: string = ''

function append (message: string) {
  text += message
}

function testWithString () {
  if (text === '') {
    append('?')

    // error TS2367: This condition will always return 'false' since the types '""' and '"?"' have no overlap.
    if (text === '?') { 
      console.log(text)
    }
  }
}
```

The same structure with a `number` instead of `string` as data type does not throw a compiler error:

```ts
let count: number = 0

function increment () {
  count += 1
}

function testWithNumber () {
  if (count === 0) {
    increment()
    if (count !== 0) { // no error here: code structure is similar, only the variable type differs
      console.log(count)
    }
  }
}
```

**Expected behavior:**

No compiler errors

**Actual behavior:**

Compiler error on the line `if (text === '?') {`

```
error TS2367: This condition will always return 'false' since the types '""' and '"?"' have no overlap.
```

**Playground Link:** 

https://www.typescriptlang.org/play?#code/DYUwLgBGIB5gXBAzmATgSwHYHMIF4IByQgKBIDMBXTAYzHQHtMIBDABzZEwBMIAKALYgkSFthCIUGHAEoIAbxIQosSAGoCQkWJAkAvmSq16TFSgDq6MAAsAymiy4+cxcvTl+0OPjwFiLpWVWDi5uPkIAfkIZQLcPPi9IXz8olwgAenSIEFRUBlQIABVbACYAZgA2AHZEQut0JAgaJm4rRmYAd3RgYFZgDpYAT0bUcEpUZkJyFmAkEEJkLBoQKGsVsEHORsIAIh2Flh4iHYj9iGsWADcVzAYIBmvUYHYAOlig5swkBlAX4AZsAlVDEghADMoDAYyKBIM1qAgIJhKAIAEY5fAQAAMhmodHaECWoyEmEgzgUgThJIgGggAEZ9DjjPjoBYrNYAHLItEFMmuAnxSlJXxYgKgwkgYlgZzvdz8QUQACEwsxaUyiLuOTyBTWo0QzW4KyklDo4xWDUWAm6LFQABp7phgINVitLtb0CwUaAoJsVq1yOQckh3spPt9fv9AYKQUFwWCGUA

**Related Issues:**

https://github.com/microsoft/TypeScript/issues/31433
https://github.com/microsoft/TypeScript/issues/28569

**Comment by MartinJohns** [comment]
Duplicate of #9998 (I almost know this number by heart).

The control-flow analysis is not aware of your modifcation within `append()`.

> The same structure with a `number` instead of `string` as data type does not throw a compiler error:

The same structure with a `number` instead of a `string` will result in the same error. But your code differs. If you align it with your `string` example you get the same issue:

    function testWithNumber () {
      if (count === 0) {
        increment()
        if (count === 1) {
          console.log(count)
        }
      }
    }

**Comment by josdejong** [comment]
Thanks for your fast reply Martin. 

You're right, with a number there is indeed the same issue after all.

In the #9998 I indeed see the exact same code structure being discussed. Looking at the amount of comments and the time it is already open I assume it is not a trivial thing to fix.

I suppose it's best to close this issue as a duplicated then?

**Comment by MartinJohns** [comment]
> I suppose it's best to close this issue as a duplicated then?

I'm not a member of the TypeScript team, I can't tell you. :-) Likely someone from the TypeScript team will just label this as a duplicate, and at some point the TypeScript bot will automatically comment and close the issue (I frequently get spam-mails due to this).

**Comment by josdejong** [comment]
👍

**Comment by MartinJohns** [comment]
@josdejong Told you. :-D

**Comment by josdejong** [comment]
😂 👍
